File: examples/example-54/ruby/features/support/messaging.rb
require 'bunny'
require './src/messaging'

messaging_start

at_exit do
    messaging_stop
end

Before('@rabbitmq') do
    @connection = Bunny.new
    @connection.start
    @channel = @connection.create_channel
    @topic_exchange = @channel.topic('topic_exchange', auto_delete: true)
    @topic_messages = []
    queue = @channel
            .queue('', exclusive: true)
            .bind(@topic_exchange, routing_key: '#')
    queue.subscribe do |delivery_info, properties, body|
        @topic_messages << { body: body, topic: delivery_info.routing_key }
        puts "---- Topic #{delivery_info.routing_key}: #{body}"
    end
end

After('@rabbitmq') { @connection.close }

File: examples/example-54/js/features/support/messaging.js
const amqp = require('amqplib');
const { Before, After, BeforeAll, AfterAll } = require('@cucumber/cucumber');
const { connectMessaging, closeMessaging } = require('../../src/messaging');

BeforeAll(async () => {
    await connectMessaging();
});

AfterAll(async () => {
    await closeMessaging();
});

Before({ tags: '@rabbitmq' }, async function () {
    const connection = await amqp.connect('amqp://localhost');
    this.amqp = { connection, topics: {} };
    this.amqp.channel = await this.amqp.connection.createChannel();

    await this.amqp.channel.assertExchange('topic_exchange', 'topic', {
        durable: false
    });

    const q = await this.amqp.channel.assertQueue('', { exclusive: true });
    await this.amqp.channel.bindQueue(q.queue, 'topic_exchange', '#');
    await this.amqp.channel.consume(q.queue, (msg) => {
        const topicName = msg.fields.routingKey;
        this.amqp.topics[topicName] = this.amqp.topics[topicName] || [];
        this.amqp.topics[topicName].push(msg.content.toString());
        console.log(`---- Topic ${topicName}: ${msg.content.toString()}`);
    }, { noAck: true });
});

After({ tags: '@rabbitmq' }, async function hook() {
    if (!this.amqp?.connection) { return; }
    await this.amqp.connection.close();
});